airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py

```diff
@@ -460,6 +465,47 @@ class ContactListStream(Stream):
     limit_field = "count"
 
 
+class DealStageHistoryStream(Stream):
+    """Deal stage history, API v1
+    Deal stage history is exposed by the v1 API, but not the v3 API.
+    The v1 endpoint requires the contacts scope.
+    Docs: https://legacydocs.hubspot.com/docs/methods/deals/get-all-deals
+    """
+    
+    url = "/deals/v1/deal/paged"
+    more_key = "hasMore"
+    data_field = "deals"
+    
+    def parse_cursor(self, record):
+        updated_at = 0
+        for prop in record.get("properties", {}).values():
+            if prop.get("timestamp", 0) > updated_at:
+                updated_at = prop.get("timestamp")
+        return updated_at
+    
+    def list(self, fields) -> Iterable:
+        params = {"propertiesWithHistory": "dealstage"}
+        yield from self.read(partial(self._api.get, url=self.url), params)
+
+
+class DealStream(CRMObjectStream):
+    """Deals, API v3"""
+    
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.stageHistory = DealStageHistoryStream(**kwargs)
+    
+    def list(self, fields) -> Iterable:
+        history_by_id = {}
+        for record in self.stageHistory.list(fields):
+            if all(field in record for field in ("dealId", "dealstage")):
+                history_by_id[record["dealId"]] = record["dealstage"]
+        for record in super().list(fields):
+            if record.get("dealId") in history_by_id:
+                record["dealstage"] = history_by_id[record.get("dealId")]
```

Comment:
```suggestion
            if record.get("id") and int(record["id"]) in history_by_id:
                record["dealstage"] = history_by_id[int(record.get("id"))]
```
Will be correct if we change code something like this. 
Incorrect record key, `deal` record has never have `dealId` key

Reply:
[The v3 endpoint uses id](https://developers.hubspot.com/docs/api/crm/deals) and [the v1 endpoint uses dealId](https://legacydocs.hubspot.com/docs/methods/deals/get-all-deals), but updated to use `id` [here](https://github.com/airbytehq/airbyte/pull/3734/files#diff-37c875e03f2e65ea1423534b0b634e18eaca797135f8d9df32360407b1d0ccb9R478) for consistency!